core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -158,14 +176,26 @@ static class ManifestReadTask extends BaseFileScanTask implements DataTask {
     @Override
     public CloseableIterable<StructLike> rows() {
       Types.NestedField readableMetricsField = projection.findField(MetricsUtil.READABLE_METRICS);
+      Types.NestedField dataSequenceNumberField =
+          projection.findField(MetadataTableUtils.DATA_SEQUENCE_NUMBER);
 
-      if (readableMetricsField == null) {
+      if (readableMetricsField == null && dataSequenceNumberField == null) {
```

Comment:
If we can reduce the number of possible branches, that would be great for code readability.

one way is to enhance withDerviedColumns to itself be smarter, and wrap conditionally.  Then here we would just return withDerivedColumns all the time.

Reply:
Now rows() method is very light and projection of computed columns moved inside the constructor of new StructLike